internal/service/serviceinstall.go

```diff
@@ -37,13 +37,33 @@ func Register(ctx *log.Context) error {
 	if !isSystemdSupported(ctx) {
 		return errors.New("Systemd not supported. Failed to register service")
 	}
+	targetVersion := os.Getenv("AZURE_GUEST_AGENT_EXTENSION_VERSION")
+	ctx.Log("message", "trying to register extension with version: "+targetVersion)
 
 	ctx.Log("message", "Generating service configuration files")
 	systemdUnitContent := generateServiceConfigurationContent(ctx)
 	serviceHandler := getSystemdHandler(ctx)
 
-	ctx.Log("message", "Registering service")
-	err := serviceHandler.Register(ctx, systemdUnitContent)
+	isInstalled, err := IsInstalled(ctx)
+	if err != nil {
+		return err
+	}
+
+	// If the service is installed, check if it needs to be upgraded.
+	if isInstalled {
+		installedVersion, err := serviceHandler.GetInstalledVersion(ctx)
+		if err != nil {
+			return err
+		}
+
+		if installedVersion == targetVersion {
```

Comment:
FYI. CRP only allows upgrades at this time for RunCommand v2. No downgrades.

Reply:
Good to know!

The reason I'm adding this command is because whenever there is an upgrade, the agent calls the `update` command for each of the previous commands executed. 

That means that if `RC_0001`, `RC_0002` and `RC_0003` were executed before, then the agent will call `update` for each of those. That may be required for normal run command (I still don't completely understand the workflow for normal run command) but for immediate run command I make this check to avoid updating the service definition for each of those requests triggered by the agent.

So, once the service definition has been updated, then I just ignore upcoming requests if they are also asking to update from X to Y.

Reply:
update is a handler level command. It is executed just once. Enable command is actually called for each extension RC_0001, RC_0002 and RC_0003. So, if there are 3 extensions, enable is called 3 times (once for each extension). The reason is because if the extension did not execute (for ex. due to update of handler, and other platform issue perhaps), it gives another chance to complete. If the extension has already reached a terminal state like Succeeded or Failed, RC_0001, RC_0002 and RC_0003 extensions won't be rerun after update.